---
title: "AF2_data_loader"
author: "Neshatul Haque"
date: '2023-03-16'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
source("~/repos/reviews/2023_AF2_conformational_survey/AF2conform_utils.R")

```

```{r}
load("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/CATHUNP2.RData")
DomDir_cth <- "/scratch/g/mtzimmermann/data/Cath_Local_db/v4_3_0"
DomDir_af2 <- "/scratch/g/mtzimmermann/data/Cath_Local_db/v4_3_0_af2_dom/"

#Data from gene ontology, kegg, etc
gok <- read.csv("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/Uniprot_try2.csv")

lf <- list.files(path = "/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/",
                 pattern = '*.tab', full.names = TRUE)
print(lf)

# ## Read in data : This data includes jessica's data as well
# d <- readr::read_tsv("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_v04.tab")
# 
# d <- d[!names(d) %in% c("af2ResBoundary", "af2DomLen", "pLDDTq10", "pLDDTq20", "pLDDTq30", 
# "RMSD", "Seq_Diff")]

# # Updataed RMSD RMSD2 and RMSD3 data
# rmsd1 <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTH_AF2_05.tab", header = T)[c(1,10)]

# pariwise alignment global-local,  no truncation
rmsd1 <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTH_AF2_10_ali_local_global_noTrunc.tab", header = T)[-2]

# # updated rmsd6 potential non-core domain gesments were removed
# rmsd6 <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/PDBUNPCTH_AF2_08.tab", header=T)

# # updated rmsd6 potential non-core domain gesments were removed with appropriate alignment
rmsd6 <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTH_AF2_09_ali_local_global.tab", header=T)#[c(1,10, 11, 12)]

# some of the colnames are changed to *_t to distinguish it from rmsd1 data
names(rmsd6) <- c("Dom", "SP_PRIMARY_t", "af2ResBoundary_t", "af2DomLen_t", "PID_t", 
"PID1_t", "pLDDTq10_", "pLDDTq20_t", "pLDDTq30_t", "RMSD6", "Nter_trunc", "Cter_trunc")

#ligand data
ligdf <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_ligand02.tab", header = T)
ligdf$Lig_Present <- NA
ligdf$Lig_Present[ligdf$Lig_Num != 0] <- 1
ligdf$Lig_Present[ligdf$Lig_Num == 0] <- 0

#metal data
metaldf <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_metal02.tab", header=T)

#Relasa data
# relasa <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_RelASA03.tab", header = T)

# relative surface area, ratio and secondary structure
relasa <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_RelASA04.tab", header = T)



#Volume and surface area of domain
volsurf <- read.table(file = file.path("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics", "VolSurf.tab"), sep = "\t", header = T)

# CATH putative domains
cth_putative <- read.table("/scratch/g/mtzimmermann/data/Cath_Local_db/cath-b-newest-all/cath-b-newest-putative")


# Seq cluster
clstrFile <- file.path("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/PDBUNPCTHAF2_SeqClstr.tab")
clstr <- read.table(clstrFile, header = T)


# Sec str content
message("Secondary structure data")
#secstr <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/PDBUNPCTH_SecStr_00.tab", header = T)
secstr <- read.table("/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics/PDBUNPCTH_SecStr_01.tab", header = T)

secstr$secstr3class <- NA
secstr$secstr3class[(secstr$cth_secstr_count - secstr$af2_secstr_count) > 0] <- "Loss"
secstr$secstr3class[(secstr$cth_secstr_count - secstr$af2_secstr_count) < 0] <- "Gain" 
secstr$secstr3class[(secstr$cth_secstr_count - secstr$af2_secstr_count) == 0] <- "NoChange"

secstr$loop3class <- NA
secstr$loop3class[(secstr$cth_loop_count - secstr$af2_loop_count) > 0] <- "Loss"
secstr$loop3class[(secstr$cth_loop_count - secstr$af2_loop_count) < 0] <- "Gain"
secstr$loop3class[(secstr$cth_loop_count - secstr$af2_loop_count) == 0] <- "NoChange"

secstr$secstr3class2 <- NA
secstr$secstr3class2[(secstr$cth_secstr_count - secstr$af2_secstr_count) > 2] <- "Loss"
secstr$secstr3class2[(secstr$cth_secstr_count - secstr$af2_secstr_count) < -2] <- "Gain" 
cond_diff2 <- ((secstr$cth_secstr_count - secstr$af2_secstr_count) <= 2) & ((secstr$cth_secstr_count - secstr$af2_secstr_count) >= -2)
secstr$secstr3class2[cond_diff2] <- "NoChange"

secstr$secstr3class4 <- NA
secstr$secstr3class4[(secstr$cth_secstr_count - secstr$af2_secstr_count) > 4] <- "Loss"
secstr$secstr3class4[(secstr$cth_secstr_count - secstr$af2_secstr_count) < -4] <- "Gain" 
cond_diff4 <- ((secstr$cth_secstr_count - secstr$af2_secstr_count) <= 4) & ((secstr$cth_secstr_count - secstr$af2_secstr_count) >= -4)
secstr$secstr3class4[cond_diff4] <- "NoChange"

secstr$secstr3class8 <- NA
secstr$secstr3class8[(secstr$cth_secstr_count - secstr$af2_secstr_count) > 8] <- "Loss"
secstr$secstr3class8[(secstr$cth_secstr_count - secstr$af2_secstr_count) < -8] <- "Gain" 
cond_diff8 <- ((secstr$cth_secstr_count - secstr$af2_secstr_count) <= 8) & ((secstr$cth_secstr_count - secstr$af2_secstr_count) >= -8)
secstr$secstr3class8[cond_diff8] <- "NoChange"

message("Secondary structure data processed")


# merge various
df1 <- merge(CATHUNP2, rmsd1, by = 'Dom')
df1 <- merge(df1, rmsd6[!names(rmsd6) %in% "SP_PRIMARY"], by = 'Dom')
df1 <- merge(df1, ligdf[!names(ligdf) %in% "SP_PRIMARY"], by = 'Dom')
df1 <- merge(df1, metaldf[!names(metaldf) %in% "SP_PRIMARY"], by = 'Dom')
df1 <- merge(df1, relasa[!names(relasa) %in% "SP_PRIMARY"], by = 'Dom')
df1 <- merge(df1, volsurf[!names(volsurf) %in% "SP_PRIMARY"], by = 'Dom')
df1 <- merge(df1, clstr, by = 'Dom')
df1 <- merge(df1, secstr, by = 'Dom')


df1$SP_C95 <- paste0(df1$SP_PRIMARY, "_", df1$C95)

df1 <- df1 %>% distinct(Dom, .keep_all = TRUE)

#remove putative domains
nPutDom <- df1 %>% filter(Dom %in% cth_putative$V1) %>% dim()

df1$Putative <- FALSE
df1[df1$Dom %in% cth_putative$V1,]$Putative <- TRUE


#RSMD normalization RMSD100 or R100
#SciViews ln natural logarithm
#log base function by default calculates natural log ln
df1$R100 <- with(df1, (RMSD/(1+log(sqrt(Dom_Len/100))))) %>% round(3)

# Mark subdomains by clustering by size
#--------------------------------------------
#'@ separates short and long domain obtained after CD HIT 95 % clustering 
#'@ differentiates between sub-domain of different length, >=20
markSubDom <- function(v, sdcol){
  # sub domain cutoff length. If two sub-domains are 20 amino acid 
  # different in length then they can be called sub-domain
  #sdcol <- 20
  # v <- z2$Dom_Len
  if(max(v) - min(v) >=sdcol){
  distance <- dist(v, method = "euclidean") # distance matrix
  fit <- hclust(distance, method="ward.D2")
  groups <- cutree(fit, h=sdcol)
  subDClust <- paste0('0', groups)
  }else{
  subDClust <- rep("00", length(v))
  }
  return(subDClust)
}


message("Clustering domains to get diversity of sub_domains within length difference of 20")
df1$SP_C95_XX <- NA
spc95IDs <- unique(df1$SP_C95)

#monitor progress
pb <- txtProgressBar(min = 1, max = length(spc95IDs), 
                     style = 3, width = 50, 
                     char = "=")

for(i in seq_along(spc95IDs)){
 # i= 1
  tempdf <- df1[df1$SP_C95 == spc95IDs[i],]
  ms <- markSubDom(tempdf$Dom_Len, 20)
  df1$SP_C95_XX[df1$SP_C95 == spc95IDs[i]] <- paste0(tempdf$SP_C95, "_", ms)
  setTxtProgressBar(pb, i)
}


message("Assigning average dom length to sub_domain SP_C95_XX")

df1$SP_C95_XX_domL <- NA
SP_C95_XX_uid <- unique(df1$SP_C95_XX)

for(i in seq_along(SP_C95_XX_uid)){
  av_dom_len <- df1$Dom_Len[df1$SP_C95_XX == SP_C95_XX_uid[i]] %>% mean() %>% round(., 2)
  df1$SP_C95_XX_domL[df1$SP_C95_XX == SP_C95_XX_uid[i]] <- av_dom_len
}



message("Clustering domains to get diversity of sub_domains within length difference of 10")
df1$SP_C95_XX10 <- NA
spc95IDs <- unique(df1$SP_C95)

#monitor progress
pb <- txtProgressBar(min = 1, max = length(spc95IDs), 
                     style = 3, width = 50, 
                     char = "=")

ms <- NULL
for(i in seq_along(spc95IDs)){
 # i= 1
  tempdf <- df1[df1$SP_C95 == spc95IDs[i],]
  ms <- markSubDom(tempdf$Dom_Len, 10)
  df1$SP_C95_XX10[df1$SP_C95 == spc95IDs[i]] <- paste0(tempdf$SP_C95, "_", ms)
  setTxtProgressBar(pb, i)
}

message("Assigning average dom length to sub_domain SP_C95_XX10")
df1$SP_C95_XX10_domL <- NA
SP_C95_XX10_uid <- unique(df1$SP_C95_XX10)

for(i in seq_along(SP_C95_XX10_uid)){
  av_dom_len <- df1$Dom_Len[df1$SP_C95_XX10 == SP_C95_XX10_uid[i]] %>% mean() %>% round(., 2)
  df1$SP_C95_XX10_domL[df1$SP_C95_XX10 == SP_C95_XX10_uid[i]] <- av_dom_len
}
#--------------------------------------------


savedf<- F
if(savedf){
  write.table(df1, "/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_08_df1.tab")
}

# Remove all the bad alignments, such as gaps in the refseq, AF2
df <- df1 %>% filter(PID1>99)
ggplot(df1, aes(RMSD, R100))+geom_point()
ggplot(df, aes(RMSD, R100))+geom_point()


savedf<- F
if(savedf){
  write.table(df, "/scratch/g/mtzimmermann/molecular_modeling/AF2_protein_dynamics//PDBUNPCTHAF2_08_df.tab")
}


## Summarize
df$fold <- paste( df$Class, df$Arch, df$Topol, sep = '.')
df$nchunks <- stringr::str_count(df$af2ResBoundary, "-")

message(stringr::str_interp( 'Total topologies represented: ${length(unique(df$fold))}' ))
message(stringr::str_interp( 'Total proteins represented  : ${length(unique(df$SP_PRIMARY))}' ))

message("resolution:")
summary(df$Str_Resol)
table(df$Str_Resol == 999)

message("Ligands:")
table(df$Lig_Num, useNA = 'a')
100 * sum(df$Lig_Num>0) / length(df$Lig_Num)

message("Metals:")
table(df$Metal_Num, useNA = 'a')
df$Metal_Num[is.na(df$Metal_Num)] <- '0'
100 * sum(df$Metal_Num != '0') / length(df$Metal_Num)

message("N.Chunks")
table(df$nchunks)
100 * sum(df$nchunks[!is.na(df$nchunks)] > 1) / dim(df)[1]

```
```{r}
# PID1_t is percent identity of the truncated domain
ggplot(df1[df1$PID1_t>99,], aes(RMSD, RMSD6)) + geom_point()

```

## Process data by clustered domain
In "df" each record is a domain before clustering. In dfSPC95XX10 each record is a cluster of domain representing one domain of a protein. Same domain of the protein might have different ID "dfSPC95XX10$ProtSpecificDomID", if the member are at least 10 residue apart. The dataframe dfSPC95XX is the same as dfSPC95XX10 but with rediue difference 20 amino acid.

```{r}
dim(df)
# 176257     76

# no of unique proteins 
df$SP_PRIMARY %>% unique() %>% length()
#[1] 3540

df$C100 %>% unique() %>% length()
#[1] 39555

# list of C95 unique member
lC95 <- df$C95 %>% unique()
df$C95 %>% unique() %>% length()
#[1] 8707

#list of SP_C95 uniwue member
lSP_C95 <- df$SP_C95 %>% unique()
df$SP_C95 %>% unique() %>% length()

#list of SP_C95_XX uniwue member
lSP_C95_XX <- df$SP_C95_XX %>% unique()
df$SP_C95_XX %>% unique() %>% length()

#list of SP_C95_XX10 uniwue member
lSP_C95_XX10 <- df$SP_C95_XX10 %>% unique()
df$SP_C95_XX10 %>% unique() %>% length()



getDomClstr2 <- function(i, clust_name){
  out <- list()
  Uni <- df$SP_PRIMARY[df[[clust_name]]==i]
  if(length(Uni!=0)){
  out$ProtSpecificDomID <- i
  
  dfselect <- df[df[[clust_name]]==i,]
  out$C95ID <- dfselect$C95[1]
  
  out$Class <- paste0(dfselect$Class, collapse = ";")
  out$Class1 <- dfselect$Class[1]
  out$nClass <- length(unique(dfselect$Class))
  
  out$UnipId <- paste0(unique(Uni), collapse = ";")
  out$nUni <- length(unique(Uni))
  

  out$Dom <- paste0(dfselect$Dom, collapse = ";")
  out$nDom <- length(dfselect$Dom)

  
  #Dom length
  DomLenlist <- dfselect$Dom_Len
  out$DomLen <- paste0(DomLenlist, collapse = ";")
  out$DomLenmin <- min(DomLenlist)
  out$DomLenmean <- mean(DomLenlist) %>% round(3)
  out$DomLenq50 <- median(DomLenlist)
  out$DomLenmax <- max(DomLenlist)
  out$DomLensd <- sd(DomLenlist) %>% round(3)
  
  #RMSD
  rmdlist <- dfselect$RMSD
  out$RMSD <- paste0(rmdlist, collapse = ";")
  out$Rmin <- min(rmdlist)
  out$Rmean <- mean(rmdlist) %>% round(3)
  out$Rq50 <- median(rmdlist)
  out$Rmax <- max(rmdlist)
  out$Rsd <- sd(rmdlist) %>% round(3)
  
  #RMSD6
  rmd6list <- dfselect$RMSD6
  out$RMSD6 <- paste0(rmd6list, collapse = ";")
  out$R6min <- min(rmd6list)
  out$R6mean <- mean(rmd6list) %>% round(3)
  out$R6q50 <- median(rmd6list)
  out$R6max <- max(rmd6list)
  out$R6sd <- sd(rmd6list) %>% round(3)
  
  #RMSD100
  r100list <- dfselect$R100
  out$R100 <- paste0(r100list, collapse = ";")
  out$R100min <- min(r100list)
  out$R100mean <- mean(r100list) %>% round(3)
  out$R100q50 <- median(r100list)
  out$R100max <- max(r100list)
  out$R100sd <- sd(r100list) %>% round(3)
  
  #cth helix
  helixlist <- dfselect$cth_Helix_count
  out$cthHelix <- paste0(helixlist, collapse = ";")
  out$cthHlxmin <- min(helixlist)
  out$cthHlxmean <- mean(helixlist) %>% round(3)
  out$cthHlxq50 <- median(helixlist)
  out$cthHlxmax <- max(helixlist)
  out$cthHlxsd <- sd(helixlist) %>% round(3)
  
  #cth sheet
  sheetlist <- dfselect$cth_Sheet_count
  out$cthSheet <- paste0(sheetlist, collapse = ";")
  out$cthShtmin <- min(sheetlist)
  out$cthShtmean <- mean(sheetlist) %>% round(3)
  out$cthShtq50 <- median(sheetlist)
  out$cthShtmax <- max(sheetlist)
  out$cthShtsd <- sd(sheetlist) %>% round(3)
  
  
  #cth loop
  looplist <- dfselect$cth_loop_count
  out$cthLoop <- paste0(looplist, collapse = ";")
  out$cthLpmin <- min(looplist)
  out$cthLpmean <- mean(looplist) %>% round(3)
  out$cthLpq50 <- median(looplist)
  out$cthLpmax <- max(looplist)
  out$cthLpsd <- sd(looplist) %>% round(3)
  
    #Area by Vol ratio
  AVlist <- with(dfselect, Area/Vol) %>% round(3)
  out$AVratio <- paste0(AVlist, collapse = ";")
  out$AVmin <- min(AVlist)
  out$AVmean <- mean(AVlist) %>% round(3)
  out$AVq50 <- median(AVlist)
  out$AVmax <- max(AVlist)
  out$AVsd <- sd(AVlist) %>% round(3)
  
   
    #secondary structure diference
  secstrlist <- with(dfselect, cth_secstr_count-af2_secstr_count)
  out$nSecStr_diff <- paste0(secstrlist, collapse = ";")
  out$nSecStr_diffmin <- min(secstrlist)
  out$nSecStr_diffmean <- mean(secstrlist) %>% round(3)
  out$nSecStr_diffq50 <- median(secstrlist)
  out$nSecStr_diffmax <- max(secstrlist)
  if(is.na(sd(secstrlist))){
    out$nSecStr_diffsd <- sd(secstrlist) %>% round(3)
  }else out$nSecStr_diffsd <- 0
  
  
    #loop diference
  looplist <- with(dfselect, cth_loop_count - af2_loop_count)
  out$nLoop_diff <- paste0(looplist, collapse = ";")
  out$nLoop_diffmin <- min(looplist)
  out$nloop_diffmean <- mean(looplist) %>% round(3)
  out$nLoop_diffq50 <- median(looplist)
  out$nLoop_diffmax <- max(looplist)
  if(!is.na(sd(looplist))){
    out$nLoop_diffsd <- sd(looplist) %>% round(3)
  }else out$nLoop_diffsd <- 0
  
    #BE ratio
  BElist <- dfselect$BE80_ratio %>% round(3)
  out$BE <- paste0(BElist, collapse = ";")
  out$BEmin <- min(BElist)
  out$BEmean <- mean(BElist) %>% round(3)
  out$BEq50 <- median(BElist)
  out$BEmax <- max(BElist)
  
  if(!is.na(sd(BElist))){
    out$BEsd <- sd(BElist) %>% round(3)
  }else out$BEsd <- 0
    #Metal 
  Metallist <- dfselect$Metal_DomResLen
  out$Met <- paste0(Metallist, collapse = ";")
  out$Metmin <- min(Metallist)
  out$Metmean <- mean(Metallist) %>% round(3)
  out$Metq50 <- median(Metallist)
  out$Metmax <- max(Metallist)
  if(!is.na(sd(Metallist))){
    out$Metsd <- sd(Metallist) %>% round(3)
  }else out$Metsd <- 0
  
  
    #lig
  Liglist <- dfselect$Lig_DomResLen
  out$Lig <- paste0(Liglist, collapse = ";")
  out$Ligmin <- min(Liglist)
  out$Ligmean <- mean(Liglist) %>% round(3)
  out$Ligq50 <- median(Liglist)
  out$Ligmax <- max(Liglist)
  if(!is.na(sd(Liglist))){
    out$Ligsd <- sd(Liglist) %>% round(3)
  }else out$Ligsd <- 0
  
  }
  return(as.data.frame(out))
}

```


```{r}

# Grouped based on protein specific subdomains
SP_C95_XX_list <- df$SP_C95_XX %>% unique()
dfSPC95XX <- do.call(rbind, pbmcapply::pbmclapply(SP_C95_XX_list, function(x){
  getDomClstr2(x, "SP_C95_XX")
  }, mc.cores = 48))

SP_C95_XX10_list <- df$SP_C95_XX10 %>% unique()
dfSPC95XX10 <- do.call(rbind, pbmcapply::pbmclapply(SP_C95_XX10_list, function(x){
  getDomClstr2(x, "SP_C95_XX10")
  }, mc.cores = 48))

```
